File: web/render.js
/*
 * render.js — draw a maze (from maze.js) onto a canvas, export PNG data URLs.
 * mazes/render.py reproduces the same geometry with PIL.
 *
 * Geometry: pitch = cellPx + wallPx. Cell (x, y) occupies the square at
 * (margin + wallPx + x*pitch, margin + wallPx + y*pitch) of size cellPx.
 * Wall slots are wallPx wide; lattice pillars are wallPx squares.
 */
(function (root, factory) {
  if (typeof module === 'object' && module.exports) { module.exports = factory(); }
  else { root.MazeRender = factory(); }
}(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  const DEFAULT_OPTIONS = {
    style: 'lines',       // lines: pillars only where walls meet | blocks: every lattice pillar filled
    cellPx: 24,
    wallPx: 4,
    margin: 12,
    theme: 'light',       // light: black walls on white | dark: white walls on black (Pathfinder-like)
    markers: 'sg',        // sg: green S / red G discs | dots: two identical discs | letters: S and G
    markerScale: 0.6,     // disc diameter as a fraction of cellPx
    grayscale: false,     // gray markers (S = filled disc, G = ring for 'sg')
    overlay: null         // null | 'solution' | 'components'
  };

  const THEMES = {
    light: { bg: '#ffffff', wall: '#000000', start: '#2ca02c', goal: '#d62728', dot: '#1f77b4', gray: '#6e6e6e' },
    dark:  { bg: '#000000', wall: '#ffffff', start: '#5ee65e', goal: '#ff6b6b', dot: '#ffd166', gray: '#9a9a9a' }
  };

  function options(opts) { return Object.assign({}, DEFAULT_OPTIONS, opts || {}); }

  function measure(maze, opts) {
    const o = options(opts);
    const pitch = o.cellPx + o.wallPx;
    return {
      width: 2 * o.margin + o.wallPx + maze.width * pitch,
      height: 2 * o.margin + o.wallPx + maze.height * pitch,
      pitch: pitch
    };
  }

  /** Wall slots: h[j][x] = wall above cell (x, j) (j = H means below the last row); v[y][i] = wall left of cell (i, y). */
  function wallSlots(maze) {
    const W = maze.width, H = maze.height;
    const h = [], v = [];
    for (let j = 0; j <= H; j++) {
      const row = [];
      for (let x = 0; x < W; x++) row.push(j === 0 || j === H || !maze.down[j - 1][x]);
      h.push(row);
    }
    for (let y = 0; y < H; y++) {
      const row = [];
      for (let i = 0; i <= W; i++) row.push(i === 0 || i === W || !maze.right[y][i - 1]);
      v.push(row);
    }
    return { h: h, v: v };
  }

  function cellOrigin(o, pitch, x, y) {
    return [o.margin + o.wallPx + x * pitch, o.margin + o.wallPx + y * pitch];
  }

  function solutionCells(maze) {
    if (maze.meta && maze.meta.solution) return maze.meta.solution;
    if (typeof MazeGen !== 'undefined') { const r = MazeGen.solve(maze); return r.path; }
    return null;
  }

  function draw(canvas, maze, opts) {
    const o = options(opts);
    const t = THEMES[o.theme] || THEMES.light;
    const m = measure(maze, o);
    const W = maze.width, H = maze.height, pitch = m.pitch;
    canvas.width = m.width;
    canvas.height = m.height;
    const ctx = canvas.getContext('2d');
    ctx.fillStyle = t.bg;
    ctx.fillRect(0, 0, m.width, m.height);

    // component tint (drawn under the walls)
    if (o.overlay === 'components' && typeof MazeGen !== 'undefined') {
      const r = MazeGen.solve(maze);
      const colors = {};
      colors[r.startComponent] = 'rgba(44,160,44,0.28)';
      colors[r.goalComponent] = 'rgba(214,39,40,0.28)';
      for (let y = 0; y < H; y++) {
        for (let x = 0; x < W; x++) {
          const c = r.components[y][x];
          ctx.fillStyle = colors[c] || 'rgba(128,128,128,0.18)';
          const p = cellOrigin(o, pitch, x, y);
          ctx.fillRect(p[0], p[1], o.cellPx, o.cellPx);
          if (x < W - 1 && maze.right[y][x]) ctx.fillRect(p[0] + o.cellPx, p[1], o.wallPx, o.cellPx);
          if (y < H - 1 && maze.down[y][x]) ctx.fillRect(p[0], p[1] + o.cellPx, o.cellPx, o.wallPx);
        }
      }
      if (!r.solvable && maze.meta && maze.meta.cutEdge) {
        // highlight the missing link
        const a = maze.meta.cutEdge[0], b = maze.meta.cutEdge[1];
        const p = cellOrigin(o, pitch, Math.min(a[0], b[0]), Math.min(a[1], b[1]));
        ctx.fillStyle = 'rgba(255,140,0,0.95)';
        if (a[1] === b[1]) ctx.fillRect(p[0] + o.cellPx - o.wallPx, p[1], 3 * o.wallPx, o.cellPx);
        else ctx.fillRect(p[0], p[1] + o.cellPx - o.wallPx, o.cellPx, 3 * o.wallPx);
      }
    }

    // walls
    const s = wallSlots(maze);
    ctx.fillStyle = t.wall;
    for (let j = 0; j <= H; j++) {
      for (let x = 0; x < W; x++) {
        if (s.h[j][x]) ctx.fillRect(o.margin + o.wallPx + x * pitch, o.margin + j * pitch, o.cellPx, o.wallPx);
      }
    }
    for (let y = 0; y < H; y++) {
      for (let i = 0; i <= W; i++) {
        if (s.v[y][i]) ctx.fillRect(o.margin + i * pitch, o.margin + o.wallPx + y * pitch, o.wallPx, o.cellPx);
      }
    }
    for (let j = 0; j <= H; j++) {
      for (let i = 0; i <= W; i++) {
        let fill = o.style === 'blocks';
        if (!fill) {
          fill = (i > 0 && s.h[j][i - 1]) || (i < W && s.h[j][i]) || (j > 0 && s.v[j - 1][i]) || (j < H && s.v[j][i]);
        }
        if (fill) ctx.fillRect(o.margin + i * pitch, o.margin + j * pitch, o.wallPx, o.wallPx);
      }
    }

    // solution overlay
    if (o.overlay === 'solution') {
      const sol = solutionCells(maze);
      if (sol && sol.length > 1) {
        ctx.strokeStyle = 'rgba(30,144,255,0.75)';
        ctx.lineWidth = Math.max(2, o.cellPx * 0.3);
        ctx.lineCap = 'round';
        ctx.lineJoin = 'round';
        ctx.beginPath();
        for (let k = 0; k < sol.length; k++) {
          const p = cellOrigin(o, pitch, sol[k][0], sol[k][1]);
          const cx = p[0] + o.cellPx / 2, cy = p[1] + o.cellPx / 2;
          if (k === 0) ctx.moveTo(cx, cy); else ctx.lineTo(cx, cy);
        }
        ctx.stroke();
      }
    }

    // markers
    const r = o.cellPx * o.markerScale / 2;
    const ps = cellOrigin(o, pitch, maze.start[0], maze.start[1]);
    const pg = cellOrigin(o, pitch, maze.goal[0], maze.goal[1]);
    const cs = [ps[0] + o.cellPx / 2, ps[1] + o.cellPx / 2];
    const cg = [pg[0] + o.cellPx / 2, pg[1] + o.cellPx / 2];
    if (o.markers === 'letters') {
      ctx.fillStyle = t.wall;
      ctx.font = 'bold ' + Math.round(o.cellPx * 0.75) + 'px sans-serif';
      ctx.textAlign = 'center';
      ctx.textBaseline = 'middle';
      ctx.fillText('S', cs[0], cs[1] + o.cellPx * 0.04);
      ctx.fillText('G', cg[0], cg[1] + o.cellPx * 0.04);
    } else if (o.markers === 'dots') {
      ctx.fillStyle = o.grayscale ? t.gray : t.dot;
      ctx.beginPath(); ctx.arc(cs[0], cs[1], r, 0, 2 * Math.PI); ctx.fill();
      ctx.beginPath(); ctx.arc(cg[0], cg[1], r, 0, 2 * Math.PI); ctx.fill();
    } else {
      ctx.fillStyle = o.grayscale ? t.gray : t.start;
      ctx.beginPath(); ctx.arc(cs[0], cs[1], r, 0, 2 * Math.PI); ctx.fill();
      if (o.grayscale) {
        ctx.strokeStyle = t.gray;
        ctx.lineWidth = Math.max(2, r * 0.45);
        ctx.beginPath(); ctx.arc(cg[0], cg[1], r - ctx.lineWidth / 2, 0, 2 * Math.PI); ctx.stroke();
      } else {
        ctx.fillStyle = t.goal;
        ctx.beginPath(); ctx.arc(cg[0], cg[1], r, 0, 2 * Math.PI); ctx.fill();
      }
    }
    return canvas;
  }

  function toDataURL(maze, opts) {
    const canvas = document.createElement('canvas');
    draw(canvas, maze, opts);
    return canvas.toDataURL('image/png');
  }

  /** Largest cellPx (>= 4) such that the image fits within maxPx. */
  function fitCellPx(maze, maxPx, opts) {
    const o = options(opts);
    const longest = Math.max(maze.width, maze.height);
    const cell = Math.floor((maxPx - 2 * o.margin - o.wallPx) / longest) - o.wallPx;
    return Math.max(4, cell);
  }

  return { DEFAULT_OPTIONS: DEFAULT_OPTIONS, THEMES: THEMES, measure: measure, draw: draw, toDataURL: toDataURL, fitCellPx: fitCellPx, wallSlots: wallSlots };
}));

File: mazes/render.py
"""Render mazes to images with PIL, using the same geometry as web/render.js.

    pitch = cell_px + wall_px
    cell (x, y) -> square at (margin + wall_px + x*pitch, margin + wall_px + y*pitch), size cell_px
    wall slots are wall_px wide; lattice pillars are wall_px squares
"""
from dataclasses import dataclass, replace
from typing import Optional

from PIL import Image, ImageDraw, ImageFont

from .generate import solve

THEMES = {
    "light": {"bg": (255, 255, 255), "wall": (0, 0, 0), "start": (44, 160, 44), "goal": (214, 39, 40), "dot": (31, 119, 180), "gray": (110, 110, 110)},
    "dark": {"bg": (0, 0, 0), "wall": (255, 255, 255), "start": (94, 230, 94), "goal": (255, 107, 107), "dot": (255, 209, 102), "gray": (154, 154, 154)},
}


@dataclass
class RenderOptions:
    style: str = "lines"         # lines | blocks
    cell_px: int = 24
    wall_px: int = 4
    margin: int = 12
    theme: str = "light"         # light | dark
    markers: str = "sg"          # sg | dots | letters
    marker_scale: float = 0.6
    grayscale: bool = False      # gray markers and an 8-bit 'L' output image
    overlay: Optional[str] = None  # None | solution | components
    supersample: int = 2         # anti-aliasing factor for discs (1 = pixel exact)

    def scaled(self, k):
        return replace(self, cell_px=self.cell_px * k, wall_px=self.wall_px * k, margin=self.margin * k, supersample=1)


def measure(maze, o: RenderOptions):
    pitch = o.cell_px + o.wall_px
    return 2 * o.margin + o.wall_px + maze.width * pitch, 2 * o.margin + o.wall_px + maze.height * pitch


def wall_slots(maze):
    """h[j][x]: wall above cell (x, j) (j == H: below last row); v[y][i]: wall left of cell (i, y)."""
    W, H = maze.width, maze.height
    h = [[j == 0 or j == H or not maze.down[j - 1][x] for x in range(W)] for j in range(H + 1)]
    v = [[i == 0 or i == W or not maze.right[y][i - 1] for i in range(W + 1)] for y in range(H)]
    return h, v


def _cell_origin(o, pitch, x, y):
    return o.margin + o.wall_px + x * pitch, o.margin + o.wall_px + y * pitch


def _rgba(rgb, alpha):
    return (rgb[0], rgb[1], rgb[2], int(round(alpha * 255)))


def _blend_rect(img, box, rgba):
    layer = Image.new("RGBA", img.size, (0, 0, 0, 0))
    ImageDraw.Draw(layer).rectangle(box, fill=rgba)
    img.alpha_composite(layer)


def _render_at(maze, o: RenderOptions):
    t = THEMES.get(o.theme, THEMES["light"])
    width, height = measure(maze, o)
    W, H = maze.width, maze.height
    pitch = o.cell_px + o.wall_px
    img = Image.new("RGBA", (width, height), t["bg"] + (255,))
    draw = ImageDraw.Draw(img)

    if o.overlay == "components":
        r = solve(maze)
        colors = {r["startComponent"]: _rgba((44, 160, 44), 0.28), r["goalComponent"]: _rgba((214, 39, 40), 0.28)}
        tint = Image.new("RGBA", img.size, (0, 0, 0, 0))
        td = ImageDraw.Draw(tint)
        for y in range(H):
            for x in range(W):
                c = colors.get(r["components"][y][x], _rgba((128, 128, 128), 0.18))
                x0, y0 = _cell_origin(o, pitch, x, y)
                td.rectangle([x0, y0, x0 + o.cell_px - 1, y0 + o.cell_px - 1], fill=c)
                if x < W - 1 and maze.right[y][x]:
                    td.rectangle([x0 + o.cell_px, y0, x0 + o.cell_px + o.wall_px - 1, y0 + o.cell_px - 1], fill=c)
                if y < H - 1 and maze.down[y][x]:
                    td.rectangle([x0, y0 + o.cell_px, x0 + o.cell_px - 1, y0 + o.cell_px + o.wall_px - 1], fill=c)
        if not r["solvable"] and maze.meta.get("cutEdge"):
            a, b = maze.meta["cutEdge"]
            x0, y0 = _cell_origin(o, pitch, min(a[0], b[0]), min(a[1], b[1]))
            if a[1] == b[1]:
                td.rectangle([x0 + o.cell_px - o.wall_px, y0, x0 + o.cell_px + 2 * o.wall_px - 1, y0 + o.cell_px - 1], fill=(255, 140, 0, 242))
            else:
                td.rectangle([x0, y0 + o.cell_px - o.wall_px, x0 + o.cell_px - 1, y0 + o.cell_px + 2 * o.wall_px - 1], fill=(255, 140, 0, 242))
        img.alpha_composite(tint)
        draw = ImageDraw.Draw(img)

    # walls
    h, v = wall_slots(maze)
    wall = t["wall"] + (255,)
    for j in range(H + 1):
        for x in range(W):
            if h[j][x]:
                x0 = o.margin + o.wall_px + x * pitch
                y0 = o.margin + j * pitch
                draw.rectangle([x0, y0, x0 + o.cell_px - 1, y0 + o.wall_px - 1], fill=wall)
    for y in range(H):
        for i in range(W + 1):
            if v[y][i]:
                x0 = o.margin + i * pitch
                y0 = o.margin + o.wall_px + y * pitch
                draw.rectangle([x0, y0, x0 + o.wall_px - 1, y0 + o.cell_px - 1], fill=wall)
    for j in range(H + 1):
        for i in range(W + 1):
            fill = o.style == "blocks"
            if not fill:
                fill = ((i > 0 and h[j][i - 1]) or (i < W and h[j][i]) or (j > 0 and v[j - 1][i]) or (j < H and v[j][i]))
            if fill:
                x0 = o.margin + i * pitch
                y0 = o.margin + j * pitch
                draw.rectangle([x0, y0, x0 + o.wall_px - 1, y0 + o.wall_px - 1], fill=wall)

    # solution overlay
    if o.overlay == "solution":
        sol = maze.meta.get("solution") or solve(maze)["path"]
        if sol and len(sol) > 1:
            pts = []
            for x, y in sol:
                x0, y0 = _cell_origin(o, pitch, x, y)
                pts.append((x0 + o.cell_px / 2, y0 + o.cell_px / 2))
            layer = Image.new("RGBA", img.size, (0, 0, 0, 0))
            ld = ImageDraw.Draw(layer)
            lw = max(2, int(round(o.cell_px * 0.3)))
            ld.line(pts, fill=(30, 144, 255, 191), width=lw, joint="curve")
            for px, py in (pts[0], pts[-1]):
                ld.ellipse([px - lw / 2, py - lw / 2, px + lw / 2, py + lw / 2], fill=(30, 144, 255, 191))
            img.alpha_composite(layer)
            draw = ImageDraw.Draw(img)

    # markers
    r = o.cell_px * o.marker_scale / 2
    sx, sy = _cell_origin(o, pitch, *maze.start)
    gx, gy = _cell_origin(o, pitch, *maze.goal)
    cs = (sx + o.cell_px / 2, sy + o.cell_px / 2)
    cg = (gx + o.cell_px / 2, gy + o.cell_px / 2)
    if o.markers == "letters":
        size = max(6, int(round(o.cell_px * 0.75)))
        try:
            font = ImageFont.truetype("DejaVuSans-Bold.ttf", size)
        except OSError:
            font = ImageFont.load_default()
        for txt, c in (("S", cs), ("G", cg)):
            draw.text(c, txt, fill=wall, font=font, anchor="mm")
    elif o.markers == "dots":
        col = (t["gray"] if o.grayscale else t["dot"]) + (255,)
        for c in (cs, cg):
            draw.ellipse([c[0] - r, c[1] - r, c[0] + r, c[1] + r], fill=col)
    else:
        col_s = (t["gray"] if o.grayscale else t["start"]) + (255,)
        draw.ellipse([cs[0] - r, cs[1] - r, cs[0] + r, cs[1] + r], fill=col_s)
        if o.grayscale:
            lw = max(2, int(round(r * 0.45)))
            draw.ellipse([cg[0] - r, cg[1] - r, cg[0] + r, cg[1] + r], outline=t["gray"] + (255,), width=lw)
        else:
            draw.ellipse([cg[0] - r, cg[1] - r, cg[0] + r, cg[1] + r], fill=t["goal"] + (255,))
    return img


def render_image(maze, options: Optional[RenderOptions] = None, **kwargs):
    """Render a maze to a PIL image ('RGB', or 'L' when grayscale)."""
    o = options or RenderOptions()
    if kwargs:
        o = replace(o, **kwargs)
    k = max(1, int(o.supersample))
    if k > 1:
        big = _render_at(maze, o.scaled(k))
        img = big.resize((big.width // k, big.height // k), Image.LANCZOS)
    else:
        img = _render_at(maze, o)
    img = img.convert("RGB")
    return img.convert("L") if o.grayscale else img


def save_png(maze, path, options: Optional[RenderOptions] = None, **kwargs):
    img = render_image(maze, options, **kwargs)
    img.save(path, format="PNG", optimize=True)
    return img
